### PR TITLE
Prevent accidental enablement of efficient attention.

### LINKED ISSUE #132004: Memory Efficient Attention on ROCm results in image corruption on the diffusers SD3 pipeline
### 🐛 Describe the bug

_(I'm happy to report this over on the diffusers repo, but I suspect it's actually a bug in the ROCm build of Pytorch 2.4.0 as you will see)_

Since Pytorch 2.4.0, I found that the HuggingFace Diffusers Stable Diffusion 3 pipeline was producing corrupted images on AMD accelerators I had access to (a weird grid pattern over the first image, followed by gibberish for subsequent images.

Example code:

```python
import torch
from diffusers import StableDiffusion3Pipeline

pipe = StableDiffusion3Pipeline.from_pretrained("stabilityai/stable-diffusion-3-medium-diffusers",torch_dtype=torch.float16)
pipe = pipe.to("cuda")

generator = torch.Generator("cuda")
generator.manual_seed(123456)

image1 = pipe(
    "A cat holding a sign that says hello world",
    negative_prompt="",
    num_inference_steps=32,
    guidance_scale=7.0,
    generator=generator
).images[0]
image2 = pipe(
    "A cat holding a sign that says hello world",
    negative_prompt="",
    num_inference_steps=32,
    guidance_scale=7.0,
    generator=generator
).images[0]
image1
```
pt24_1

```python
image2
```

pt24_2

Generating further images from the pipeline generates more gibberish.  I have reproduced this on AMD Mi300x, AMD Mi250 and AMD Mi210.

On experimentation I have determined that this is due to the addition of memory efficient attention in the ROCm build of PyTorch 2.4.0 because disabling this `torch.backends.cuda.enable_mem_efficient_sdp(False)` produces "correct" images.

pt24_1_nomemeffatt
pt24_2_nomemeffatt

This problem **does not** occur on:

1. Pytorch 2.3.1 for ROCm (which does not have memory efficient attention compiled in, as per the warning it prints).
2. Pytorch 2.4.0 on Nvidia GPUs (tested A100)

### Versions

On the Mi250 box (I no longer have access to the Mi300x system, which was running AlmaLinux 9.4 with the latest install of ROCm 6):

```
Collecting environment information...
PyTorch version: 2.4.0+rocm6.0
Is debug build: False
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: 6.0.32830-d62f6a171

OS: Rocky Linux release 8.9 (Green Obsidian) (x86_64)
GCC version: (GCC) 15.0.0 20240724 (experimental)
Clang version: Could not collect
CMake version: version 3.26.5
Libc version: glibc-2.28

Python version: 3.11.7 (main, Jan 16 2024, 03:27:23) [GCC 10.2.0] (64-bit runtime)
Python platform: Linux-4.18.0-513.24.1.el8_9.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: AMD Instinct MI250X/MI250 (gfx90a:sramecc+:xnack-)
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: 6.0.32830
MIOpen runtime version: 3.0.0
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              128
On-line CPU(s) list: 0-127
Thread(s) per core:  1
Core(s) per socket:  64
Socket(s):           2
NUMA node(s):        8
Vendor ID:           AuthenticAMD
CPU family:          25
Model:               1
Model name:          AMD EPYC 7763 64-Core Processor
Stepping:            1
CPU MHz:             2450.000
CPU max MHz:         3529.0520
CPU min MHz:         1500.0000
BogoMIPS:            4900.05
Virtualization:      AMD-V
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            32768K
NUMA node0 CPU(s):   0-15
NUMA node1 CPU(s):   16-31
NUMA node2 CPU(s):   32-47
NUMA node3 CPU(s):   48-63
NUMA node4 CPU(s):   64-79
NUMA node5 CPU(s):   80-95
NUMA node6 CPU(s):   96-111
NUMA node7 CPU(s):   112-127
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd amd_ppin brs arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca

Versions of relevant libraries:
[pip3] numpy==1.26.3
[pip3] pytorch-triton-rocm==3.0.0
[pip3] torch==2.4.0+rocm6.0
[pip3] torchaudio==2.4.0+rocm6.0
[pip3] torchvision==0.19.0+rocm6.0
[conda] Could not collect
```

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### PR DESCRIPTION
**Pull request by xinyazhang**
Currently Efficient attention and Flash attention share the same set of GPU
kernels on ROCM and have common limitations on head sizes.

Fixes #132004

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mcarilli @ptrblck @leslie-fang-intel @jgong5 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn

### CODE REVIEW

**Comment by jithunnair-amd** [comment]
@xinyazhang Can you please confirm whether the `test_transformers.py::TestSDPAFailureModesCUDA::test_invalid_fused_inputs_head_dim_kernel1_cuda` failure in ROCm CI job for this PR is unrelated to this PR?

**Comment by xinyazhang** [comment]
> Can you please confirm whether the `test_transformers.py::TestSDPAFailureModesCUDA::test_invalid_fused_inputs_head_dim_kernel1_cuda` failure in ROCm CI job for this PR is unrelated to this PR?

It seems the UT's have different assumptions on Efficient attention and we need to change the UT accordingly.
I'll go through `test_transformer.py` locally and label keep-going to ensure we didn't miss anything.

File: test/test_transformers.py [hunk 1]
```diff
@@ -1493,6 +1495,8 @@ def test_invalid_fused_inputs_attn_mask_present(self, device, kernel: SDPBackend
     @onlyCUDA
     @unittest.skipIf(not PLATFORM_SUPPORTS_FLASH_ATTENTION, "Does not support fused SDPA or pre-SM80 hardware")
     def test_unaligned_tensors(self, device):
+        if TEST_WITH_ROCM:  # ROCm GPU kernels does not have strict alignment requirements
+            return
```
Comment (malfet) [inline test/test_transformers.py:1499]: This comment seems contradictory to what it actually does, feels like you want to skip test with ROCM
```suggestion
```

File: test/test_transformers.py [hunk 2]
```diff
@@ -1493,6 +1495,8 @@ def test_invalid_fused_inputs_attn_mask_present(self, device, kernel: SDPBackend
     @onlyCUDA
     @unittest.skipIf(not PLATFORM_SUPPORTS_FLASH_ATTENTION, "Does not support fused SDPA or pre-SM80 hardware")
```
Comment (malfet) [inline test/test_transformers.py:1496]:
```suggestion
    @unittest.skipIf(not PLATFORM_SUPPORTS_FLASH_ATTENTION, "Does not support fused SDPA or pre-SM80 hardware")
    @unittest.skipIf(TEST_WITH_ROCM, "ROCM GPU kernels have strict alignment requirement")
```

**Review by malfet — APPROVED** [review]
LGTM, though marked `test_unaligned_tensors` as skipped, let me know if you disagree

**Review by malfet — CHANGES_REQUESTED** [review]
This should go into main, shouldn't it? Tried to auto-rebase but it did not work
@jithunnair-amd can you please clarify why PR targets 2.4 rather than main? Were there a more generic change in trunk already?

File: test/test_transformers.py (hunk 1 above)
Comment (xinyazhang) [reply test/test_transformers.py:1499]: This UT acts in an inverted manner: it expects a RuntimeError thrown for irregular shapes to pass, otherwise fails.

However ROCm's implementation does not have such restrictions and no RuntimeError will be thrown, and thus the UT failed because the inputs are acceptable on ROCm.

**Comment by xinyazhang** [comment]
> This should go into main, shouldn't it? Tried to auto-rebase but it did not work

Moved the target to main

**Comment by xinyazhang** [comment]
Re-creating the PR
